apps/vault-apy/src/APYSimulator.ts

```diff
@@ -563,4 +655,101 @@ export default class APYSimulator {
       throw new Error('Save to db failed');
     }
   }
+
+  async getPeriodFinishForVault(vaultAddress: string, forkBlock: number): Promise<number> {
+    // Find the vault data in the config
+    const vaultData = this.#config.vaults.find(v => v.address.toLowerCase() === vaultAddress.toLowerCase());
+    
+    if (!vaultData) {
+      throw new Error(`Vault address ${vaultAddress} not found in config`);
+    }
+
+    if (!vaultData.gauge) {
+      throw new Error(`Gauge address not specified for vault ${vaultAddress}`);
+    }
+
+    const provider = new ethers.providers.JsonRpcProvider(this.#config.alchemyUrl);
+
+    // ABI fragment for the periodFinish variable
+    const gaugeABI = [
+      "function periodFinish() view returns (uint256)"
+    ];
+
+    const gaugeContract = new Contract(vaultData.gauge, gaugeABI, provider);
+
+    try {
+      // Call the periodFinish function at the specified block
+      const periodFinish: BigNumber = await gaugeContract.callStatic.periodFinish({
+        blockTag: forkBlock
+      });
+      log(periodFinish.toString());
+      return periodFinish.toNumber();
+    } catch (error) {
+      console.error(`Error fetching periodFinish for gauge ${vaultData.gauge} at block ${forkBlock}:`, error);
+      throw error;
+    }
+  }
+
+  // Add this helper method to the APYSimulator class
+  #getMostRecentThursdayMidnight(timestamp: number): number {
+    const date = new Date(timestamp * 1000);
+    const day = date.getUTCDay();
+    const daysToSubtract = (day + 3) % 7; // Thursday is 4, so we add 3 and take modulo 7
+    date.setUTCDate(date.getUTCDate() - daysToSubtract);
+    date.setUTCHours(0, 0, 0, 0);
+    return Math.floor(date.getTime() / 1000);
+  }
+
+  #getNextThursdayMidnightUTC(timestamp: number): number {
+    const date = new Date(timestamp * 1000);
+    const daysUntilThursday = (4 - date.getUTCDay() + 7) % 7;
+    date.setUTCDate(date.getUTCDate() + daysUntilThursday);
+    date.setUTCHours(0, 0, 0, 0);
+    return Math.floor(date.getTime() / 1000);
+  }
+
+  async runAll() {
+    if (!this.#config.vaults.length) {
+      log('Skipping, no vaults specified');
+      return;
+    }
+
+    const latestBlock = await this.#alchemyProvider.getBlockNumber();
+    log(`Processing all vaults at block ${latestBlock}`);
+
+    for (const vault of this.#config.vaults) {
+      log(`Processing vault: ${vault.address}`);
+      try {
+        let forkBlock = latestBlock;
```

Comment:
Make your adjustments inside functions as much as possible, it will make maintaining this code a lot easier. If there are a slight differences in these adjustments with the others, try to fit it into logic groupings so that future adjustments can be done easily without much chance of breaking other functionality.